[ZIP 401] Addressing mempool denial-of-service (#275)
zcash/zips: drafts/zip-daira-mempool-limitation.rst

::

>   ZIP: Unassigned

Reviewer: 400 series

Author: ```suggestion
  ZIP: 401
```

>   Title: Addressing mempool denial-of-service

Reviewer: ```suggestion
  Title: Addressing mempool denial-of-service in zcashd
```

  Owners: Daira Hopwood <daira@electriccoin.co>
  Status: Draft
  Category: Network
  Created: 2019-09-09
  License: MIT

Terminology
===========

The key words "MUST", "SHOULD", and "MAY" in this document are to be interpreted
as described in RFC 2119. [#RFC2119]_

Abstract
========

This proposal specifies a change to the behaviour of zcashd nodes intended to
mitigate denial-of-service from transaction flooding.

Motivation
==========

Adoption of this proposal would increase robustness of Zcash nodes against
denial-of-service attack, in particular attacks that attempt to exhaust node
memory.

Bitcoin Core added size limitation for the mempool in version 0.12

> [#BitcoinCore-PR6722]_, defaulting to 300 MB. This was after Zcash forked from

Reviewer: ```suggestion
[#BitcoinCore-PR6722]_, defaulting to 300 MB. This was after ``zcashd`` forked from
```

> Bitcoin Core.

Reviewer: ```suggestion
Bitcoin Core, so ``zcashd`` did not inherit this change.
```

i.e. an adversary should not be able to cause legitimate transactions (that do not
themselves present any denial-of-service problem) to be preferentially evicted
relative to its own transactions.

Any configuration options should have reasonable defaults, i.e. without changing
relevant configuration, a node should be adequately protected from denial-of-service
via mempool memory exhaustion.

Non-requirements
================

The current architecture of Zcash imposes fundamental limits on scaling of
transaction throughput. This proposal does not increase the aggregate transaction
capacity of the network. (The Blossom upgrade does increase transaction capacity,
by a factor of two [#zip-0208]_.)

Denial-of-service issues in the messaging layer of the peer-to-peer protocol are
out of scope for this proposal.

This proposal is focused primarily on memory exhaustion attacks. It does not
attempt to use fees to make denial-of-service economically prohibitive, since that
is unlikely to be feasible while maintaining low fees for legitimate users. It
does not preclude changes in fee policy.

Specification
=============

> This specification describes the intended behaviour of zcashd nodes. Other node

Reviewer: ```suggestion
This specification describes the intended behaviour of ``zcashd`` nodes. Other node
```

implementations MAY implement the same or similar behaviour, but this is not a
requirement of the network protocol. Thus, RFC 2119 conformance keywords below are

> to be interpreted only as placing requirements on the zcashd implementation (and

Reviewer: ```suggestion
to be interpreted only as placing requirements on the ``zcashd`` implementation (and
```

potentially other implementations that have adopted this specification in full).

The mempool of a node holds a set of transactions. Each transaction has a cost,
which is an integer defined as:

  max(serialized transaction size in bytes, 4000) + fee_penalty

> where ``fee_penalty`` is 16000 if the transaction pays a fee less than

Reviewer 2: The `fee_penalty` may have the unintended side effect of encouraging an attacker to use transactions with less than the standard fee if they are trying to fill up the mempool, because this makes it cheaper and require fewer transactions to do so. Suggestion: Do not take the `fee_penalty` in to account when calculating the total cost of the mempool.

10000 zatoshi, otherwise 0.

Each node also MUST hold a FIFO queue RecentlyEvicted of pairs (txid, time), where
the time indicates when the given txid was evicted. This SHOULD be empty on node

> startup. The size of RecentlyEvicted MUST never exceed 10000 entries.

Reviewer: ```suggestion
startup. The size of RecentlyEvicted SHOULD never exceed 100000 entries.
```

Reviewer: Other node implementations might manage this queue in ways that do not cause memory pressure.

Reviewer 3: Hmm I think maybe it does - so is the maximum effective size of the RecentlyEvicted list is the nominal size divided by the number of equivalent transactions that can be created using the same funds but different transaction IDs? Without the attacker risking the transactions being mined?

Author: Is this now explained adequately in the rationale section?

> There MUST be a configuration option ``mempool.tx_cost_limit``, which SHOULD default

Reviewer: ```suggestion
There MUST be a configuration option ``mempool.total_cost_limit``, which SHOULD default
```

to 80000000.

There MUST be a configuration option ``mempool.eviction_memory_minutes``, which
SHOULD default to 60.

On receiving a transaction:

* If it is in RecentlyEvicted, the transaction MUST be dropped.
* Calculate its cost. If the total cost of transactions in the mempool including

>   this one would exceed ``mempool.tx_cost_limit``, then the node MUST repeatedly

Reviewer: ```suggestion
  this one would exceed ``mempool.total_cost_limit``, then the node MUST repeatedly
```

  call EvictTransaction (with the new transaction included as a candidate to evict)

>   until the total cost does not exceed ``mempool.tx_cost_limit``.

Reviewer: ```suggestion
  until the total cost does not exceed ``mempool.total_cost_limit``.
```

EvictTransaction MUST do the following:

> * Select a random transaction to evict, weighted by cost.

Reviewer 3: *weighted linearly?

* Add the txid and the current time to RecentlyEvicted, dropping the oldest entry

>   in RecentlyEvicted if necessary to keep it to at most 10000 entries.

Reviewer: ```suggestion
  in RecentlyEvicted if necessary to keep it to at most 100000 entries.
```

and/or just before RecentlyEvicted is accessed when receiving a transaction.

Rationale
=========

The accounting for transaction size should include some overhead per transaction,
to reflect the cost to the network of processing them (proof and signature
verification; networking overheads; size of in-memory data structures). The
implication of not including overhead is that a denial-of-service attacker would
be likely to use minimum-size transactions so that more of them would fit in a
block, increasing the unaccounted-for overhead. A possible counterargument would
be that the complexity of accounting for this overhead is unwarranted given that
the format of a transaction already imposes a minimum size. However, the proposed
cost function is almost as simple as using transaction size directly.

The threshold 4000 for the cost function is chosen so that the size in bytes of a
typical fully shielded Sapling transaction (with, say, 2 shielded outputs and up
to 5 shielded inputs) will fall below the threshold. This has the effect of
ensuring that such transactions are not evicted preferentially to typical

> transparent transactions because of their size.

Reviewer: We know of at least two legitimate use patterns that substantially diverge from this: 

1) Exchanges sweeping from user-specific ingress addresses to other parts of their fund management infrastructure will likely have a lot of inputs from consolidating small notes

2) Miner payouts will have lots of outputs

However, it seems reasonable to address the problem we have now based on the behavior we see now. Since this is not a consensus-critical ZIP it will be simple to address future problems that may arise.

The proposed eviction policy differs significantly from that of Bitcoin Core
[#BitcoinCore-PR6722]_, which is primarily fee-based. This reflects differing
philosophies about the motivation for fees and the level of fee that legitimate
users can reasonably be expected to pay. The proposed cost function does involve
a penalty for transactions with a fee lower than the standard (0.0001 ZEC) value,
but since there is no further benefit to increasing the fee above the standard
value, it creates no pressure toward escalating fees. For transactions up to
4000 bytes, this penalty makes a transaction that pays less than the standard fee

> value five times as likely to be chosen for eviction.

Reviewer: ```suggestion
value five times as likely to be chosen for eviction (because 4000 + 16000 = 20000 = 4000 \* 5).
```

> The default value of 80000000 for ``mempool.tx_cost_limit`` represents no more

Reviewer: ```suggestion
The default value of 80000000 for ``mempool.total_cost_limit`` represents no more
```

> than 40 blocks’ worth of transactions in the worst case, which is the default

Reviewer: Consider expanding this explanation.

expiration height after Blossom network upgrade [#zip-0208]_. It would serve no
purpose to make it larger.

> The ``mempool.tx_cost_limit`` is a per-node configurable parameter in order to

Reviewer: ```suggestion
The ``mempool.total_cost_limit`` is a per-node configurable parameter in order to
```

provide flexibility for node operators to change it either in response to
attempted denial-of-service attacks, or if needed to handle spikes in transaction
demand. It may also be useful for nodes running in memory-constrained environments
to reduce this parameter.

The limit of 100000 entries in RecentlyEvicted bounds the memory needed for this
data structure. Since a txid is 32 bytes and a timestamp 8 bytes, 100000 entries
can be stored in ~4 MB, which is small compared to other node memory usage (in
particular, small compared to the maximum memory usage of the mempool itself under

> the default ``mempool.tx_cost_limit``). 100000 entries should be sufficient to

Reviewer: ```suggestion
the default ``mempool.total_cost_limit``). 100000 entries should be sufficient to
```

mitigate any performance loss caused by re-accepting transactions that were 

> previously evicted. In particular, since a transaction has a minimum cost of 1000,

Reviewer: ```suggestion
previously evicted. In particular, since a transaction has a minimum cost of 4000,
```

> and the default ``mempool.tx_cost_limit`` is 80000000, at most 80000 transactions

Reviewer: ```suggestion
and the default ``mempool.total_cost_limit`` is 80000000, at most 80000 transactions
```

Reviewer: 80M / 4000 = 20000
